## Final Project Submission

Please fill out:
* Student name: GRACE KOKI MUTUKU
* Student pace: FULL TIME
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


In [27]:
# Import the necessary libraries
import pandas as pd
import numpy as np
import sqlite3
import zipfile
import os

path = r"D:\MORINGA\MoringaGit\dsc-data-science-env-config\Phase_1\Week 3\dsc-phase-1-project-v2-4\zippedData"

In [28]:
os.chdir(path)
os.listdir()

['bom.movie_gross.csv.gz',
 'im.db',
 'im.db.zip',
 'rt.movie_info.tsv.gz',
 'rt.reviews.tsv.gz',
 'tmdb.movies.csv.gz',
 'tn.movie_budgets.csv.gz']

In [29]:
with zipfile.ZipFile('im.db.zip') as f:
    f.extractall()    

In [30]:
# conn = sqlite3.connect("im.db")
# df = pd.read_sql("""SELECT name FROM sqlite_master WHERE type = 'table';""", conn)
# df

In [31]:
# q = """
# SELECT *
#   FROM writers
# ;
# # """
# pd.read_sql(q, conn).head()

In [32]:
# q = """
# SELECT *
#   FROM persons
# ;
# """
# pd.read_sql(q, conn).head()

In [33]:
# q = """
# SELECT *
#   FROM principals
# ;
# """
# pd.read_sql(q, conn).head()

In [34]:
# q = """
# SELECT *
#   FROM known_for
# ;
# """
# pd.read_sql(q, conn).head()

In [35]:
# q = """
# SELECT *
#   FROM movie_basics
# ;
# """
# pd.read_sql(q, conn).head()

In [36]:
# q = """
# SELECT *
#   FROM movie_ratings
# ;
# """
# pd.read_sql(q, conn).head()

In [37]:
# q = """
# SELECT *
#   FROM movie_akas
# ;
# """
# pd.read_sql(q, conn).head()

## Importing Data

The `read_csv()` function is used to import all the csv files into Pandas DataFrame then the proper encoding of `latin-1` is applied.

In [38]:
# Import all the '.csv' file using a proper encoding
movies_df = pd.read_csv('tmdb.movies.csv.gz', encoding='latin-1', index_col= 0)
movie_gross_df = pd.read_csv('bom.movie_gross.csv.gz', encoding='latin-1')
movie_budgets_df = pd.read_csv('tn.movie_budgets.csv.gz', encoding='latin-1', index_col= 0)
movie_info_df = pd.read_csv('rt.movie_info.tsv.gz', delimiter='\t', encoding='latin-1')
reviews_df = pd.read_csv('rt.reviews.tsv.gz', delimiter= '\t', encoding='latin-1')

In [39]:
#movies_df 
# movie_gross_df
# movie_budgets_df
# movie_info_df 
# reviews_df
# movies_df['vote_count'].mode()

In [40]:
def dataframe_structure(df):
    print(f'The shape of the dataframe is: {df.shape}\n')
    df.info()
    display(df.describe())
    print(f'Checking for Null Values: \n',(df.isna().sum()))
    display(df.head())

## 1 Movies_df

In [41]:
dataframe_structure(movies_df)

The shape of the dataframe is: (26517, 9)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26517 entries, 0 to 26516
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   genre_ids          26517 non-null  object 
 1   id                 26517 non-null  int64  
 2   original_language  26517 non-null  object 
 3   original_title     26517 non-null  object 
 4   popularity         26517 non-null  float64
 5   release_date       26517 non-null  object 
 6   title              26517 non-null  object 
 7   vote_average       26517 non-null  float64
 8   vote_count         26517 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 2.0+ MB


id    popularity  vote_average    vote_count
count   26517.000000  26517.000000  26517.000000  26517.000000
mean   295050.153260      3.130912      5.991281    194.224837
std    153661.615648      4.355229      1.852946    960.961095
min        27.000000      0.600000      0.000000      1.000000
25%    157851.000000      0.600000      5.000000      2.000000
50%    309581.000000      1.374000      6.000000      5.000000
75%    419542.000000      3.694000      7.000000     28.000000
max    608444.000000     80.773000     10.000000  22186.000000

Checking for Null Values: 
 genre_ids            0
id                   0
original_language    0
original_title       0
popularity           0
release_date         0
title                0
vote_average         0
vote_count           0
dtype: int64


genre_ids     id original_language  \
0      [12, 14, 10751]  12444                en   
1  [14, 12, 16, 10751]  10191                en   
2        [12, 28, 878]  10138                en   
3      [16, 35, 10751]    862                en   
4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

In [42]:
movies_df.isna().sum()

genre_ids            0
id                   0
original_language    0
original_title       0
popularity           0
release_date         0
title                0
vote_average         0
vote_count           0
dtype: int64

There seems to be no missing values in the `tmdb.movies.csv.gz` file since there are 26517 entries in all columns and also checking for the sum of `NaN` values returs zero in every column confirming the absence of null values.

The next step will be to check for the presence of duplicate entries in the `movie_df` based on the subset `id` which is assumed to be unique. 1020 duplicates were identified and removed.

In [43]:
duplicates = movies_df[movies_df.duplicated(subset = 'id')] # Identify duplicate rows based on id
movies_df_no_duplicates = movies_df.drop_duplicates() # Drop the duplicate rows
movies_df_no_duplicates.head()

genre_ids     id original_language  \
0      [12, 14, 10751]  12444                en   
1  [14, 12, 16, 10751]  10191                en   
2        [12, 28, 878]  10138                en   
3      [16, 35, 10751]    862                en   
4        [28, 878, 12]  27205                en   

                                 original_title  popularity release_date  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533   2010-11-19   
1                      How to Train Your Dragon      28.734   2010-03-26   
2                                    Iron Man 2      28.515   2010-05-07   
3                                     Toy Story      28.005   1995-11-22   
4                                     Inception      27.920   2010-07-16   

                                          title  vote_average  vote_count  
0  Harry Potter and the Deathly Hallows: Part 1           7.7       10788  
1                      How to Train Your Dragon           7.7        7610  
2                                    Iron Man 2           6.8       12368  
3                                     Toy Story           7.9       10174  
4                                     Inception           8.3       22186

In [44]:
for column in movies_df.columns:
    print(column, '\n', movies_df_no_duplicates[column].value_counts(normalize = True), '\n\n')

genre_ids 
 [99]                   0.139820
[]                     0.096521
[18]                   0.083108
[35]                   0.063615
[27]                   0.044123
                         ...   
[28, 14, 35]           0.000039
[27, 53, 18, 10402]    0.000039
[28, 18, 53, 12]       0.000039
[10749, 16, 14]        0.000039
[28, 12, 18, 9648]     0.000039
Name: genre_ids, Length: 2477, dtype: float64 


id 
 71677     0.000039
478883    0.000039
105945    0.000039
298459    0.000039
414455    0.000039
            ...   
302150    0.000039
234567    0.000039
451657    0.000039
64586     0.000039
524288    0.000039
Name: id, Length: 25497, dtype: float64 


original_language 
 en    0.877907
fr    0.018943
es    0.017218
ru    0.011570
ja    0.009570
        ...   
kk    0.000039
yi    0.000039
ha    0.000039
hy    0.000039
dz    0.000039
Name: original_language, Length: 76, dtype: float64 


original_title 
 Home                                        0.000235
Lucky               

In [45]:
# movies_df_no_duplicates.loc[movies_df_no_duplicates['genre_ids'] == '[]', :]
# movies_df[movies_df.duplicated(subset = 'title')]

## 2 Movie Gross df

In [46]:
dataframe_structure(movie_gross_df)

The shape of the dataframe is: (3387, 5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


domestic_gross         year
count    3.359000e+03  3387.000000
mean     2.874585e+07  2013.958075
std      6.698250e+07     2.478141
min      1.000000e+02  2010.000000
25%      1.200000e+05  2012.000000
50%      1.400000e+06  2014.000000
75%      2.790000e+07  2016.000000
max      9.367000e+08  2018.000000

Checking for Null Values: 
 title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64


title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [47]:
# movie_gross_df.tail(800)

## Understanding the data
There are 5 columns and 3387 entries in the `bom.movie_gross.csv` file. `studio`, `domestic_gross` and `foreign_gross` have missing values. There seems to be no `NaN` values in the file.

The columns have appropriate datatypes apart from `foreign_gross` which should be a `float` value and `year` which should be type `date`.

In [48]:
movie_gross_df['foreign_gross'] = movie_gross_df['foreign_gross'].str.replace(',', '') # Replace the ',' with an empty string, so as to be able to convert to float
movie_gross_df['foreign_gross'] = movie_gross_df['foreign_gross'].astype(float) # Convert the 'foreign_gross' column datatype from object to a float

In [49]:
movie_gross_df['year'] = pd.to_datetime(movie_gross_df['year'], format = '%Y') # Convert the datatype of 'year' from int to datetime


In [50]:
movie_gross_df.describe().round(1)

domestic_gross  foreign_gross
count          3359.0         2037.0
mean       28745845.1     74872810.2
std        66982498.2    137410600.8
min             100.0          600.0
25%          120000.0      3700000.0
50%         1400000.0     18700000.0
75%        27900000.0     74900000.0
max       936700000.0    960500000.0

## Cleaning of the data
1. Check for duplicates

In [51]:
duplicates = movie_gross_df[movie_gross_df.duplicated()] # Checking for duplicate entries
len(duplicates)

0

There are no duplicate rows.

In [52]:
movie_gross_df.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

Since 39% of the values in `foreign_gross` are null, it is preferable to drop the entire column rather than replace the missing values or drop the rows with missing values and distort data of loose invaluable data respectively. 

In [53]:
movie_gross_df = movie_gross_df.drop('foreign_gross', axis = 1) # Drop the 'foreign_gross column' 
movie_gross_df

title      studio  domestic_gross  \
0                                     Toy Story 3          BV     415000000.0   
1                      Alice in Wonderland (2010)          BV     334200000.0   
2     Harry Potter and the Deathly Hallows Part 1          WB     296000000.0   
3                                       Inception          WB     292600000.0   
4                             Shrek Forever After        P/DW     238700000.0   
...                                           ...         ...             ...   
3382                                    The Quake       Magn.          6200.0   
3383                  Edward II (2018 re-release)          FM          4800.0   
3384                                     El Pacto        Sony          2500.0   
3385                                     The Swan  Synergetic          2400.0   
3386                            An Actor Prepares       Grav.          1700.0   

           year  
0    2010-01-01  
1    2010-01-01  
2    2010-01-01  
3    2010-01-01  
4    2010-01-01  
...         ...  
3382 2018-01-01  
3383 2018-01-01  
3384 2018-01-01  
3385 2018-01-01  
3386 2018-01-01  

[3387 rows x 4 columns]

In [54]:
movie_gross_df.loc[movie_gross_df['domestic_gross'].isnull() == True, : ] #Checking the entries with null values in the 'domestic_gross' column
clean_movie_gross_df = movie_gross_df.dropna(axis = 0) # Dropping all rows with null values
clean_movie_gross_df.describe().round(1)

domestic_gross
count          3356.0
mean       28771489.6
std        67006943.1
min             100.0
25%          120000.0
50%         1400000.0
75%        27950000.0
max       936700000.0

## 3 Movie Budgets
Moving on to the next file


In [55]:
dataframe_structure(movie_budgets_df)

The shape of the dataframe is: (5782, 5)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5782 entries, 1 to 82
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   release_date       5782 non-null   object
 1   movie              5782 non-null   object
 2   production_budget  5782 non-null   object
 3   domestic_gross     5782 non-null   object
 4   worldwide_gross    5782 non-null   object
dtypes: object(5)
memory usage: 271.0+ KB


release_date      movie production_budget domestic_gross  \
count           5782       5782              5782           5782   
unique          2418       5698               509           5164   
top     Dec 31, 2014  King Kong       $20,000,000             $0   
freq              24          3               231            548   

       worldwide_gross  
count             5782  
unique            5356  
top                 $0  
freq               367

Checking for Null Values: 
 release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
dtype: int64


release_date                                        movie  \
id                                                              
1   Dec 18, 2009                                       Avatar   
2   May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
3    Jun 7, 2019                                 Dark Phoenix   
4    May 1, 2015                      Avengers: Age of Ultron   
5   Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

   production_budget domestic_gross worldwide_gross  
id                                                   
1       $425,000,000   $760,507,625  $2,776,345,279  
2       $410,600,000   $241,063,875  $1,045,663,875  
3       $350,000,000    $42,762,350    $149,762,350  
4       $330,600,000   $459,005,868  $1,403,013,963  
5       $317,000,000   $620,181,382  $1,316,721,747

## Understanding the data
There are 5 columns and 5782 entries in `tn.movie_budgets.csv` file and `NaN` values seem to be absent .

The datatype of `release_date` should be converted to `datetime`, and the datatypes of `production_budget`, `domestic_gross`, and `worldwide_gross` should be converted to `int`.

In [56]:
movie_budgets_df['release_date'] = pd.to_datetime(movie_budgets_df['release_date']) # Converting the 'release_date' to datetime datatype

In [57]:
movie_budgets_df

release_date                                        movie  \
id                                                             
1    2009-12-18                                       Avatar   
2    2011-05-20  Pirates of the Caribbean: On Stranger Tides   
3    2019-06-07                                 Dark Phoenix   
4    2015-05-01                      Avengers: Age of Ultron   
5    2017-12-15            Star Wars Ep. VIII: The Last Jedi   
..          ...                                          ...   
78   2018-12-31                                       Red 11   
79   1999-04-02                                    Following   
80   2005-07-13                Return to the Land of Wonders   
81   2015-09-29                         A Plague So Pleasant   
82   2005-08-05                            My Date With Drew   

   production_budget domestic_gross worldwide_gross  
id                                                   
1       $425,000,000   $760,507,625  $2,776,345,279  
2       $410,600,000   $241,063,875  $1,045,663,875  
3       $350,000,000    $42,762,350    $149,762,350  
4       $330,600,000   $459,005,868  $1,403,013,963  
5       $317,000,000   $620,181,382  $1,316,721,747  
..               ...            ...             ...  
78            $7,000             $0              $0  
79            $6,000        $48,482        $240,495  
80            $5,000         $1,338          $1,338  
81            $1,400             $0              $0  
82            $1,100       $181,041        $181,041  

[5782 rows x 5 columns]

The `production_budget`, `domestic_gross`, and `worldwide_gross` columns are converted from a string values to a numeric values. Before the conversion, the string values are first stripped off any non-numeric characters using the `replace()` function.

In [58]:
movie_budgets_df = movie_budgets_df.assign(production_budget = pd.to_numeric(movie_budgets_df['production_budget'].str.replace('[^0-9.]', '')))
movie_budgets_df = movie_budgets_df.assign(domestic_gross = pd.to_numeric(movie_budgets_df['domestic_gross'].str.replace('[^0-9.]', '')))
movie_budgets_df = movie_budgets_df.assign(worldwide_gross = pd.to_numeric(movie_budgets_df['worldwide_gross'].str.replace('[^0-9.]', '')))

## Cleaning the data

First check for duplicates. 
We find out that there are no duplicate entries in `movie_budgets_df`.

In [59]:
duplicates = movie_budgets_df[movie_budgets_df.duplicated()] # Checking for duplicate entries
len(duplicates)

0

In [60]:
movie_budgets_df.describe().round(0)

production_budget  domestic_gross  worldwide_gross
count             5782.0          5782.0     5.782000e+03
mean          31587757.0      41873327.0     9.148746e+07
std           41812077.0      68240597.0     1.747200e+08
min               1100.0             0.0     0.000000e+00
25%            5000000.0       1429534.0     4.125415e+06
50%           17000000.0      17225945.0     2.798445e+07
75%           40000000.0      52348662.0     9.764584e+07
max          425000000.0     936662225.0     2.776345e+09

In [61]:
for column in movie_budgets_df.columns: # Checking for any other data descrepancies or extraneous values
    print(column, '\n', movie_budgets_df[column].value_counts(normalize = True), '\n\n')

release_date 
 2014-12-31    0.004151
2015-12-31    0.003978
2010-12-31    0.002594
2008-12-31    0.002421
2013-12-31    0.002248
                ...   
2018-04-27    0.000173
1982-09-17    0.000173
2010-02-10    0.000173
1966-07-06    0.000173
1977-07-13    0.000173
Name: release_date, Length: 2418, dtype: float64 


movie 
 King Kong                             0.000519
Home                                  0.000519
Halloween                             0.000519
Night of the Living Dead              0.000346
Pete’s Dragon                    0.000346
                                        ...   
Enter the Void                        0.000173
The Emperor's Club                    0.000173
Batman v Superman: Dawn of Justice    0.000173
Doubt                                 0.000173
Walking to Paris                      0.000173
Name: movie, Length: 5698, dtype: float64 


production_budget 
 20000000    0.039952
10000000    0.036666
30000000    0.030612
15000000    0.029920
250000


## 4th File

In [62]:
dataframe_structure(movie_info_df)

The shape of the dataframe is: (1560, 12)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1560 entries, 0 to 1559
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            1560 non-null   int64 
 1   synopsis      1498 non-null   object
 2   rating        1557 non-null   object
 3   genre         1552 non-null   object
 4   director      1361 non-null   object
 5   writer        1111 non-null   object
 6   theater_date  1201 non-null   object
 7   dvd_date      1201 non-null   object
 8   currency      340 non-null    object
 9   box_office    340 non-null    object
 10  runtime       1530 non-null   object
 11  studio        494 non-null    object
dtypes: int64(1), object(11)
memory usage: 146.4+ KB


id
count  1560.000000
mean   1007.303846
std     579.164527
min       1.000000
25%     504.750000
50%    1007.500000
75%    1503.250000
max    2000.000000

Checking for Null Values: 
 id                 0
synopsis          62
rating             3
genre              8
director         199
writer           449
theater_date     359
dvd_date         359
currency        1220
box_office      1220
runtime           30
studio          1066
dtype: int64


id                                           synopsis rating  \
0   1  This gritty, fast-paced, and innovative police...      R   
1   3  New York City, not-too-distant-future: Eric Pa...      R   
2   5  Illeana Douglas delivers a superb performance ...      R   
3   6  Michael Douglas runs afoul of a treacherous su...      R   
4   7                                                NaN     NR   

                                 genre          director  \
0  Action and Adventure|Classics|Drama  William Friedkin   
1    Drama|Science Fiction and Fantasy  David Cronenberg   
2    Drama|Musical and Performing Arts    Allison Anders   
3           Drama|Mystery and Suspense    Barry Levinson   
4                        Drama|Romance    Rodney Bennett   

                            writer  theater_date      dvd_date currency  \
0                   Ernest Tidyman   Oct 9, 1971  Sep 25, 2001      NaN   
1     David Cronenberg|Don DeLillo  Aug 17, 2012   Jan 1, 2013        $   
2                   Allison Anders  Sep 13, 1996  Apr 18, 2000      NaN   
3  Paul Attanasio|Michael Crichton   Dec 9, 1994  Aug 27, 1997      NaN   
4                     Giles Cooper           NaN           NaN      NaN   

  box_office      runtime             studio  
0        NaN  104 minutes                NaN  
1    600,000  108 minutes  Entertainment One  
2        NaN  116 minutes                NaN  
3        NaN  128 minutes                NaN  
4        NaN  200 minutes                NaN

## Understanding the data
There are 12 columns and 1560 entries in `rt.movie_info.tsv` file and all the columns have string values apart from `id` column which has integer values. All the columns apart form the `id` column have `NaN` values.

The datatype of `theater_date` and `dvd_date` should be converted to `datetime`, and the datatypes of `box_office` and `runtime` should be converted to `int` values.

## Cleaning the data

1. Check for duplicates and remove them. There are no duplicates in the `movie_info_df`.
2. Deal with null values
3. Changing column types.



In [63]:
duplicates = movie_info_df[movie_info_df.duplicated()] # Checking for duplicate entries
len(duplicates)

0

In [64]:
movie_info_df['theater_date'] = pd.to_datetime(movie_info_df['theater_date']) # Converting the 'theater_date' to datetime datatype
movie_info_df['dvd_date'] = pd.to_datetime(movie_info_df['dvd_date']) # Converting the 'dvd_date' to datetime datatype

In [65]:
movie_info_df = movie_info_df.assign(box_office = pd.to_numeric(movie_info_df['box_office'].str.replace('[^0-9.]', '')))
movie_info_df = movie_info_df.assign(runtime = pd.to_numeric(movie_info_df['runtime'].str.replace('[^0-9.]', '')))

In [66]:
movie_info_df

id                                           synopsis rating  \
0        1  This gritty, fast-paced, and innovative police...      R   
1        3  New York City, not-too-distant-future: Eric Pa...      R   
2        5  Illeana Douglas delivers a superb performance ...      R   
3        6  Michael Douglas runs afoul of a treacherous su...      R   
4        7                                                NaN     NR   
...    ...                                                ...    ...   
1555  1996  Forget terrorists or hijackers -- there's a ha...      R   
1556  1997  The popular Saturday Night Live sketch was exp...     PG   
1557  1998  Based on a novel by Richard Powell, when the l...      G   
1558  1999  The Sandlot is a coming-of-age story about a g...     PG   
1559  2000  Suspended from the force, Paris cop Hubert is ...      R   

                                                  genre            director  \
0                   Action and Adventure|Classics|Drama    William Friedkin   
1                     Drama|Science Fiction and Fantasy    David Cronenberg   
2                     Drama|Musical and Performing Arts      Allison Anders   
3                            Drama|Mystery and Suspense      Barry Levinson   
4                                         Drama|Romance      Rodney Bennett   
...                                                 ...                 ...   
1555   Action and Adventure|Horror|Mystery and Suspense                 NaN   
1556                 Comedy|Science Fiction and Fantasy        Steve Barron   
1557  Classics|Comedy|Drama|Musical and Performing Arts      Gordon Douglas   
1558    Comedy|Drama|Kids and Family|Sports and Fitness  David Mickey Evans   
1559  Action and Adventure|Art House and Internation...                 NaN   

                                                writer theater_date  \
0                                       Ernest Tidyman   1971-10-09   
1                         David Cronenberg|Don DeLillo   2012-08-17   
2                                       Allison Anders   1996-09-13   
3                      Paul Attanasio|Michael Crichton   1994-12-09   
4                                         Giles Cooper          NaT   
...                                                ...          ...   
1555                                               NaN   2006-08-18   
1556  Terry Turner|Tom Davis|Dan Aykroyd|Bonnie Turner   1993-07-23   
1557                                               NaN   1962-01-01   
1558                  David Mickey Evans|Robert Gunter   1993-04-01   
1559                                        Luc Besson   2001-09-27   

       dvd_date currency  box_office  runtime             studio  
0    2001-09-25      NaN         NaN    104.0                NaN  
1    2013-01-01        $    600000.0    108.0  Entertainment One  
2    2000-04-18      NaN         NaN    116.0                NaN  
3    1997-08-27      NaN         NaN    128.0                NaN  
4           NaT      NaN         NaN    200.0                NaN  
...         ...      ...         ...      ...                ...  
1555 2007-01-02        $  33886034.0    106.0    New Line Cinema  
1556 2001-04-17      NaN         NaN     88.0  Paramount Vantage  
1557 2004-05-11      NaN         NaN    111.0                NaN  
1558 2002-01-29      NaN         NaN    101.0                NaN  
1559 2003-02-11      NaN         NaN     94.0  Columbia Pictures  

[1560 rows x 12 columns]


## 5th File

In [67]:
dataframe_structure(reviews_df)

The shape of the dataframe is: (54432, 8)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54432 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54432 non-null  int64 
 1   review      48869 non-null  object
 2   rating      40915 non-null  object
 3   fresh       54432 non-null  object
 4   critic      51710 non-null  object
 5   top_critic  54432 non-null  int64 
 6   publisher   54123 non-null  object
 7   date        54432 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.3+ MB


id    top_critic
count  54432.000000  54432.000000
mean    1045.706882      0.240594
std      586.657046      0.427448
min        3.000000      0.000000
25%      542.000000      0.000000
50%     1083.000000      0.000000
75%     1541.000000      0.000000
max     2000.000000      1.000000

Checking for Null Values: 
 id                0
review         5563
rating        13517
fresh             0
critic         2722
top_critic        0
publisher       309
date              0
dtype: int64


id                                             review rating   fresh  \
0   3  A distinctly gallows take on contemporary fina...    3/5   fresh   
1   3  It's an allegory in search of a meaning that n...    NaN  rotten   
2   3  ... life lived in a bubble in financial dealin...    NaN   fresh   
3   3  Continuing along a line introduced in last yea...    NaN   fresh   
4   3             ... a perverse twist on neorealism...     NaN   fresh   

           critic  top_critic         publisher               date  
0      PJ Nabarro           0   Patrick Nabarro  November 10, 2018  
1  Annalee Newitz           0           io9.com       May 23, 2018  
2    Sean Axmaker           0  Stream on Demand    January 4, 2018  
3   Daniel Kasman           0              MUBI  November 16, 2017  
4             NaN           0      Cinema Scope   October 12, 2017


## Understanding the data
There are 8 columns and 54432 entries in `rt.reviews.tsv` file and all the columns have string values apart from `id` and `top_critic` which have integer values. The following columns have `NaN` values: `review`, `rating`, `critic` and `publisher`.

The datatype of values in `date` column should be converted to `datetime`

## Cleaning the data

1. Check for duplicates and remove them. There are 9 duplicate entries in `movie_info_df` are identified and removed.
2. Convert the `date` column's datatype to datetime.
3. Deal with null values
   - The `rating` column is dropped because it has 25% of its entries being null. It's data is closely linked to the `fresh`    column which has no null values.
   - The null values in `review`, `critic` and `publisher` columns will be replaced with `No Description`, `No Name` and `Missing Publisher Name` respectively. These columns and their entries have been kept because they have invaluable data.

In [68]:
duplicates = reviews_df[reviews_df.duplicated()] # Checking for duplicate entries
reviews_df_no_duplicates = reviews_df.drop_duplicates() # Drop the duplicate rows

In [69]:
reviews_df['date'] = pd.to_datetime(reviews_df['date']) # Converting the 'date' column values to datetime datatype

In [70]:
dataframe_structure(reviews_df_no_duplicates)

The shape of the dataframe is: (54423, 8)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54423 entries, 0 to 54431
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54423 non-null  int64 
 1   review      48867 non-null  object
 2   rating      40907 non-null  object
 3   fresh       54423 non-null  object
 4   critic      51710 non-null  object
 5   top_critic  54423 non-null  int64 
 6   publisher   54114 non-null  object
 7   date        54423 non-null  object
dtypes: int64(2), object(6)
memory usage: 3.7+ MB


id    top_critic
count  54423.000000  54423.000000
mean    1045.668596      0.240634
std      586.661865      0.427472
min        3.000000      0.000000
25%      542.000000      0.000000
50%     1083.000000      0.000000
75%     1541.000000      0.000000
max     2000.000000      1.000000

Checking for Null Values: 
 id                0
review         5556
rating        13516
fresh             0
critic         2713
top_critic        0
publisher       309
date              0
dtype: int64


id                                             review rating   fresh  \
0   3  A distinctly gallows take on contemporary fina...    3/5   fresh   
1   3  It's an allegory in search of a meaning that n...    NaN  rotten   
2   3  ... life lived in a bubble in financial dealin...    NaN   fresh   
3   3  Continuing along a line introduced in last yea...    NaN   fresh   
4   3             ... a perverse twist on neorealism...     NaN   fresh   

           critic  top_critic         publisher               date  
0      PJ Nabarro           0   Patrick Nabarro  November 10, 2018  
1  Annalee Newitz           0           io9.com       May 23, 2018  
2    Sean Axmaker           0  Stream on Demand    January 4, 2018  
3   Daniel Kasman           0              MUBI  November 16, 2017  
4             NaN           0      Cinema Scope   October 12, 2017

In [71]:
reviews_df_no_duplicates = reviews_df_no_duplicates.drop('rating', axis = 1) # Drop the 'rating' column 

In [72]:
reviews_df_no_duplicates

id                                             review   fresh  \
0         3  A distinctly gallows take on contemporary fina...   fresh   
1         3  It's an allegory in search of a meaning that n...  rotten   
2         3  ... life lived in a bubble in financial dealin...   fresh   
3         3  Continuing along a line introduced in last yea...   fresh   
4         3             ... a perverse twist on neorealism...    fresh   
...     ...                                                ...     ...   
54427  2000  The real charm of this trifle is the deadpan c...   fresh   
54428  2000                                                NaN  rotten   
54429  2000                                                NaN  rotten   
54430  2000                                                NaN  rotten   
54431  2000                                                NaN   fresh   

                  critic  top_critic         publisher                date  
0             PJ Nabarro           0   Patrick Nabarro   November 10, 2018  
1         Annalee Newitz           0           io9.com        May 23, 2018  
2           Sean Axmaker           0  Stream on Demand     January 4, 2018  
3          Daniel Kasman           0              MUBI   November 16, 2017  
4                    NaN           0      Cinema Scope    October 12, 2017  
...                  ...         ...               ...                 ...  
54427      Laura Sinagra           1     Village Voice  September 24, 2002  
54428  Michael Szymanski           0        Zap2it.com  September 21, 2005  
54429       Emanuel Levy           0   EmanuelLevy.Com       July 17, 2005  
54430   Christopher Null           0    Filmcritic.com   September 7, 2003  
54431    Nicolas Lacroix           0      Showbizz.net   November 12, 2002  

[54423 rows x 7 columns]

In [73]:
reviews_df_no_duplicates.loc[reviews_df_no_duplicates['review'].isnull() == True, 'review'] = 'No Description'
reviews_df_no_duplicates.loc[reviews_df_no_duplicates['critic'].isnull() == True, 'critic'] = 'No Name'
reviews_df_no_duplicates.loc[reviews_df_no_duplicates['publisher'].isnull() == True, 'publisher'] = 'No Publisher Name'

In [74]:
reviews_df_no_duplicates.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 54423 entries, 0 to 54431
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   id          54423 non-null  int64 
 1   review      54423 non-null  object
 2   fresh       54423 non-null  object
 3   critic      54423 non-null  object
 4   top_critic  54423 non-null  int64 
 5   publisher   54423 non-null  object
 6   date        54423 non-null  object
dtypes: int64(2), object(5)
memory usage: 3.3+ MB


In [75]:
reviews_df_no_duplicates

id                                             review   fresh  \
0         3  A distinctly gallows take on contemporary fina...   fresh   
1         3  It's an allegory in search of a meaning that n...  rotten   
2         3  ... life lived in a bubble in financial dealin...   fresh   
3         3  Continuing along a line introduced in last yea...   fresh   
4         3             ... a perverse twist on neorealism...    fresh   
...     ...                                                ...     ...   
54427  2000  The real charm of this trifle is the deadpan c...   fresh   
54428  2000                                     No Description  rotten   
54429  2000                                     No Description  rotten   
54430  2000                                     No Description  rotten   
54431  2000                                     No Description   fresh   

                  critic  top_critic         publisher                date  
0             PJ Nabarro           0   Patrick Nabarro   November 10, 2018  
1         Annalee Newitz           0           io9.com        May 23, 2018  
2           Sean Axmaker           0  Stream on Demand     January 4, 2018  
3          Daniel Kasman           0              MUBI   November 16, 2017  
4                No Name           0      Cinema Scope    October 12, 2017  
...                  ...         ...               ...                 ...  
54427      Laura Sinagra           1     Village Voice  September 24, 2002  
54428  Michael Szymanski           0        Zap2it.com  September 21, 2005  
54429       Emanuel Levy           0   EmanuelLevy.Com       July 17, 2005  
54430   Christopher Null           0    Filmcritic.com   September 7, 2003  
54431    Nicolas Lacroix           0      Showbizz.net   November 12, 2002  

[54423 rows x 7 columns]

In [81]:
# joined_df = pd.merge(movies_df_no_duplicates, movie_budgets_df, left_on = 'title', right_on ='movie', how='outer')
merged_df = pd.merge(joined_df, clean_movie_gross_df, on = 'title', how='outer')
# joined_df.head()
merged_df.head()

genre_ids       id original_language  \
0      [12, 14, 10751]  12444.0                en   
1  [14, 12, 16, 10751]  10191.0                en   
2        [12, 28, 878]  10138.0                en   
3      [16, 35, 10751]    862.0                en   
4        [28, 878, 12]  27205.0                en   

                                 original_title  popularity release_date_x  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533     2010-11-19   
1                      How to Train Your Dragon      28.734     2010-03-26   
2                                    Iron Man 2      28.515     2010-05-07   
3                                     Toy Story      28.005     1995-11-22   
4                                     Inception      27.920     2010-07-16   

                                          title  vote_average  vote_count  \
0  Harry Potter and the Deathly Hallows: Part 1           7.7     10788.0   
1                      How to Train Your Dragon           7.7      7610.0   
2                                    Iron Man 2           6.8     12368.0   
3                                     Toy Story           7.9     10174.0   
4                                     Inception           8.3     22186.0   

  release_date_y                     movie  production_budget  \
0            NaT                       NaN                NaN   
1     2010-03-26  How to Train Your Dragon        165000000.0   
2     2010-05-07                Iron Man 2        170000000.0   
3     1995-11-22                 Toy Story         30000000.0   
4     2010-07-16                 Inception        160000000.0   

   domestic_gross_x  worldwide_gross studio  domestic_gross_y       year  
0               NaN              NaN    NaN               NaN        NaT  
1       217581232.0      494870992.0   P/DW       217600000.0 2010-01-01  
2       312433331.0      621156389.0   Par.       312400000.0 2010-01-01  
3       191796233.0      364545516.0    NaN               NaN        NaT  
4       292576195.0      835524642.0     WB       292600000.0 2010-01-01

In [83]:
dataframe_structure(merged_df)

The shape of the dataframe is: (30460, 17)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30460 entries, 0 to 30459
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   genre_ids          25582 non-null  object        
 1   id                 25582 non-null  float64       
 2   original_language  25582 non-null  object        
 3   original_title     25582 non-null  object        
 4   popularity         25582 non-null  float64       
 5   release_date_x     25582 non-null  object        
 6   title              26655 non-null  object        
 7   vote_average       25582 non-null  float64       
 8   vote_count         25582 non-null  float64       
 9   release_date_y     6016 non-null   datetime64[ns]
 10  movie              6016 non-null   object        
 11  production_budget  6016 non-null   float64       
 12  domestic_gross_x   6016 non-null   float64       
 13  worldwide_gross  

id    popularity  vote_average    vote_count  \
count   25582.000000  25582.000000  25582.000000  25582.000000   
mean   293980.992417      3.068009      5.979239    184.171175   
std    154719.489871      4.317203      1.864311    932.464523   
min        27.000000      0.600000      0.000000      1.000000   
25%    154206.250000      0.600000      5.000000      1.000000   
50%    306949.500000      1.323000      6.000000      5.000000   
75%    420412.250000      3.513750      7.000000     26.000000   
max    608444.000000     80.773000     10.000000  22186.000000   

       production_budget  domestic_gross_x  worldwide_gross  domestic_gross_y  
count       6.016000e+03      6.016000e+03     6.016000e+03      3.522000e+03  
mean        3.156118e+07      4.208852e+07     9.208997e+07      2.924441e+07  
std         4.195904e+07      6.946835e+07     1.769194e+08      6.794843e+07  
min         1.100000e+03      0.000000e+00     0.000000e+00      1.000000e+02  
25%         5.000000e+06      1.294304e+06     4.000000e+06      1.230000e+05  
50%         1.700000e+07      1.705731e+07     2.821572e+07      1.500000e+06  
75%         4.000000e+07      5.207568e+07     9.698609e+07      2.910000e+07  
max         4.250000e+08      9.366622e+08     2.776345e+09      9.367000e+08

Checking for Null Values: 
 genre_ids             4878
id                    4878
original_language     4878
original_title        4878
popularity            4878
release_date_x        4878
title                 3805
vote_average          4878
vote_count            4878
release_date_y       24444
movie                24444
production_budget    24444
domestic_gross_x     24444
worldwide_gross      24444
studio               26938
domestic_gross_y     26938
year                 26938
dtype: int64


genre_ids       id original_language  \
0      [12, 14, 10751]  12444.0                en   
1  [14, 12, 16, 10751]  10191.0                en   
2        [12, 28, 878]  10138.0                en   
3      [16, 35, 10751]    862.0                en   
4        [28, 878, 12]  27205.0                en   

                                 original_title  popularity release_date_x  \
0  Harry Potter and the Deathly Hallows: Part 1      33.533     2010-11-19   
1                      How to Train Your Dragon      28.734     2010-03-26   
2                                    Iron Man 2      28.515     2010-05-07   
3                                     Toy Story      28.005     1995-11-22   
4                                     Inception      27.920     2010-07-16   

                                          title  vote_average  vote_count  \
0  Harry Potter and the Deathly Hallows: Part 1           7.7     10788.0   
1                      How to Train Your Dragon           7.7      7610.0   
2                                    Iron Man 2           6.8     12368.0   
3                                     Toy Story           7.9     10174.0   
4                                     Inception           8.3     22186.0   

  release_date_y                     movie  production_budget  \
0            NaT                       NaN                NaN   
1     2010-03-26  How to Train Your Dragon        165000000.0   
2     2010-05-07                Iron Man 2        170000000.0   
3     1995-11-22                 Toy Story         30000000.0   
4     2010-07-16                 Inception        160000000.0   

   domestic_gross_x  worldwide_gross studio  domestic_gross_y       year  
0               NaN              NaN    NaN               NaN        NaT  
1       217581232.0      494870992.0   P/DW       217600000.0 2010-01-01  
2       312433331.0      621156389.0   Par.       312400000.0 2010-01-01  
3       191796233.0      364545516.0    NaN               NaN        NaT  
4       292576195.0      835524642.0     WB       292600000.0 2010-01-01

Drop irrelevant columns

In [97]:
dataset = merged_df.drop(['id', 'original_title', 'title','movie', 'year', 'domestic_gross_y'], axis = 1)
dataframe_structure(dataset)

The shape of the dataframe is: (30460, 11)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30460 entries, 0 to 30459
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   genre_ids          25582 non-null  object        
 1   original_language  25582 non-null  object        
 2   popularity         25582 non-null  float64       
 3   release_date_x     25582 non-null  object        
 4   vote_average       25582 non-null  float64       
 5   vote_count         25582 non-null  float64       
 6   release_date_y     6016 non-null   datetime64[ns]
 7   production_budget  6016 non-null   float64       
 8   domestic_gross_x   6016 non-null   float64       
 9   worldwide_gross    6016 non-null   float64       
 10  studio             3522 non-null   object        
dtypes: datetime64[ns](1), float64(6), object(4)
memory usage: 4.0+ MB


popularity  vote_average    vote_count  production_budget  \
count  25582.000000  25582.000000  25582.000000       6.016000e+03   
mean       3.068009      5.979239    184.171175       3.156118e+07   
std        4.317203      1.864311    932.464523       4.195904e+07   
min        0.600000      0.000000      1.000000       1.100000e+03   
25%        0.600000      5.000000      1.000000       5.000000e+06   
50%        1.323000      6.000000      5.000000       1.700000e+07   
75%        3.513750      7.000000     26.000000       4.000000e+07   
max       80.773000     10.000000  22186.000000       4.250000e+08   

       domestic_gross_x  worldwide_gross  
count      6.016000e+03     6.016000e+03  
mean       4.208852e+07     9.208997e+07  
std        6.946835e+07     1.769194e+08  
min        0.000000e+00     0.000000e+00  
25%        1.294304e+06     4.000000e+06  
50%        1.705731e+07     2.821572e+07  
75%        5.207568e+07     9.698609e+07  
max        9.366622e+08     2.776345e+09

Checking for Null Values: 
 genre_ids             4878
original_language     4878
popularity            4878
release_date_x        4878
vote_average          4878
vote_count            4878
release_date_y       24444
production_budget    24444
domestic_gross_x     24444
worldwide_gross      24444
studio               26938
dtype: int64


genre_ids original_language  popularity release_date_x  \
0      [12, 14, 10751]                en      33.533     2010-11-19   
1  [14, 12, 16, 10751]                en      28.734     2010-03-26   
2        [12, 28, 878]                en      28.515     2010-05-07   
3      [16, 35, 10751]                en      28.005     1995-11-22   
4        [28, 878, 12]                en      27.920     2010-07-16   

   vote_average  vote_count release_date_y  production_budget  \
0           7.7     10788.0            NaT                NaN   
1           7.7      7610.0     2010-03-26        165000000.0   
2           6.8     12368.0     2010-05-07        170000000.0   
3           7.9     10174.0     1995-11-22         30000000.0   
4           8.3     22186.0     2010-07-16        160000000.0   

   domestic_gross_x  worldwide_gross studio  
0               NaN              NaN    NaN  
1       217581232.0      494870992.0   P/DW  
2       312433331.0      621156389.0   Par.  
3       191796233.0      364545516.0    NaN  
4       292576195.0      835524642.0     WB

In [98]:
dataset.loc[dataset['studio'].isnull() == True, 'studio'] = 'No Studio Name'


In [99]:
dataset.dropna(axis = 0).head() # Dropping all rows with null values

genre_ids original_language  popularity release_date_x  \
1  [14, 12, 16, 10751]                en      28.734     2010-03-26   
2        [12, 28, 878]                en      28.515     2010-05-07   
3      [16, 35, 10751]                en      28.005     1995-11-22   
4        [28, 878, 12]                en      27.920     2010-07-16   
5      [12, 14, 10751]                en      26.691     2010-02-11   

   vote_average  vote_count release_date_y  production_budget  \
1           7.7      7610.0     2010-03-26        165000000.0   
2           6.8     12368.0     2010-05-07        170000000.0   
3           7.9     10174.0     1995-11-22         30000000.0   
4           8.3     22186.0     2010-07-16        160000000.0   
5           6.1      4229.0     2010-02-12         95000000.0   

   domestic_gross_x  worldwide_gross          studio  
1       217581232.0      494870992.0            P/DW  
2       312433331.0      621156389.0            Par.  
3       191796233.0      364545516.0  No Studio Name  
4       292576195.0      835524642.0              WB  
5        88768303.0      223050874.0  No Studio Name